test/FunctionalTests/UnaryMethodTests.cs

```diff
@@ -164,5 +164,31 @@ public async Task MessageSentInMultipleChunks_SuccessResponse()
             var responseMessage = MessageHelpers.AssertReadMessage<HelloReply>(await response.Content.ReadAsByteArrayAsync().DefaultTimeout());
             Assert.AreEqual("Hello World", responseMessage.Message);
         }
+
+        [Test]
+        public async Task SendHeadersTwice_ThrowsException()
+        {
+            // Arrange
+            var requestMessage = new HelloRequest
+            {
+                Name = "World"
+            };
+
+            var ms = new MemoryStream();
+            MessageHelpers.WriteMessage(ms, requestMessage);
+
+            // Act
+            var response = await Fixture.Client.PostAsync(
+                "Greet.Greeter/SayHelloSendHeadersTwice",
+                new StreamContent(ms)).DefaultTimeout();
+
+            // Assert
+            Assert.AreEqual(HttpStatusCode.OK, response.StatusCode);
+            Assert.AreEqual("identity", response.Headers.GetValues("grpc-encoding").Single());
+            Assert.AreEqual("application/grpc", response.Content.Headers.ContentType.MediaType);
+
+            var responseMessage = MessageHelpers.AssertReadMessage<HelloReply>(await response.Content.ReadAsByteArrayAsync().DefaultTimeout());
+            Assert.AreEqual("Exception validated", responseMessage.Message);
```

Comment:
In general, I find relying on magic values set in a completely different place in code (the service impl in this case) not very readable.  One approach that I found useful when writing tests from Grpc.Core was to define a "universal" service that allows setting implementations of serverside handlers directly in the test.
Then many tests look like a serverside handler that's provided as a lambda at the start of the test in the Arrange phase.   When reading the test, one can easily see what corresponding serverside handler is doing and the test becomes much more readable.
Another advantage is that you don't need to come up with complicated serverside handler implementations that would be good enough for use in many different test cases.

See example here: 
https://github.com/grpc/grpc/blob/f4079fe185a51967e2773587c7804bd329936a16/src/csharp/Grpc.Core.Tests/ClientServerTest.cs#L73

A side note: for some tests, the test service uses a custom marshaller and the requests and responses are not protobuf messages, but strings. This makes the test code a bit more compact in some cases.

Reply:
This is basically just a suggestion - fine to address later, but I think it might be worthwhile to at least consider.

Reply:
Interesting idea. It is a bit more difficult for our functional tests because we simulate building the entire ASP.NET Core application.

I *think* I could emulate something like this. Will look at it later

Reply:
Filed https://github.com/grpc/grpc-dotnet/issues/80 to track